awx/main/tasks.py

```diff
@@ -520,18 +520,47 @@ def update_inventory_computed_fields(inventory_id, should_update_hosts=True):
     Signal handler and wrapper around inventory.update_computed_fields to
     prevent unnecessary recursive calls.
     '''
-    i = Inventory.objects.filter(id=inventory_id)
-    if not i.exists():
-        logger.error("Update Inventory Computed Fields failed due to missing inventory: " + str(inventory_id))
-        return
-    i = i[0]
-    try:
-        i.update_computed_fields(update_hosts=should_update_hosts)
-    except DatabaseError as e:
-        if 'did not affect any rows' in str(e):
-            logger.debug('Exiting duplicate update_inventory_computed_fields task.')
+    def _perform_update(inventory_id, should_update_hosts=True):
+        i = Inventory.objects.filter(id=inventory_id)
+        if not i.exists():
+            logger.error("Update Inventory Computed Fields failed due to missing inventory: " + str(inventory_id))
             return
-        raise
+        i = i[0]
+        try:
+            i.update_computed_fields(update_hosts=should_update_hosts)
+        except DatabaseError as e:
+            if 'did not affect any rows' in str(e):
+                logger.debug('Exiting duplicate update_inventory_computed_fields task.')
+                return
+            raise
+
+    compute_key = 'inventory_compute_{}'.format(inventory_id)
+    wait_key = 'inventory_wait_{}'.format(inventory_id)
+    with advisory_lock(compute_key, wait=False) as compute_lock:
+        if compute_lock:
+            _perform_update(inventory_id, should_update_hosts=True)
+        else:
+            with advisory_lock(wait_key, wait=False) as wait_lock:
```

Comment:
Whenever I see nested locks like this, it _immediately_ makes me wary of the potential for a deadlock:

`compute_lock` and resource `wait_lock` are used by process X and process Y

- X starts to use `compute_lock` and calls `_perform_update`.
- Around the same time, X and Y try to start using `wait_lock`
- Y 'wins' and gets `wait_lock` first
- now Y needs to use `compute_lock`
- `compute_lock` is locked by X, which is waiting for Y

Reply:
There must be something that you know that I don't.

With the `wait=False` it wouldn't happen on this line, unless I fundamentally have misunderstood the function front-to-back. No matter what, _this_ is not blocking. There is only 1 blocking lock acquisition, and that's the one holding the wait lock, waiting for the compute lock to be released. In _any_ other case, the execution continues.

> Around the same time, X and Y try to start using wait_lock
> Y 'wins' and gets wait_lock first

X bails, since it failed to obtain the wait lock

> now Y needs to use compute_lock

Yes, Y is blocking on this

> compute_lock is locked by X, which is waiting for Y

X never did anything. It finished because neither lock was obtained. It does nothing else.

Reply:
Hopefully it behaves that way 😄.  Nested locking code like this is _notorious_ for having weird edge cases - especially as you involve multiple parallel processes.  It's easy for bugs and deadlocks to sneak in, and then you're awake at 2AM trying to understand why your processes are hung 😉 

I feel like this code is sufficiently complicated to a point that it's not worth whatever improved performance it may provide.